src/lib/jobrunner.rs

```diff
@@ -0,0 +1,409 @@
+                            stderrout_file,
+                            stderr_hup: false,
+                            stdout_hup: false,
+                        });
+                    }
+                }
+            }
+        }
+
+        Err(Some(qj))
+    }
+
+    /// After a job has been inserted / removed from `self.running`, this function must be called
+    /// so that `poll()` is called with upto date file descriptors.
+    fn update_pollfds(&mut self) {
+        for (i, jobslot) in self.running.iter().enumerate() {
+            let (stderr_fd, stdout_fd) = if let Some(job) = jobslot {
+                let stderr_fd = if job.stderr_hup {
+                    -1
+                } else {
+                    job.child.stderr.as_ref().unwrap().as_raw_fd()
+                };
+                let stdout_fd = if job.stdout_hup {
+                    -1
+                } else {
+                    job.child.stdout.as_ref().unwrap().as_raw_fd()
+                };
+                (stderr_fd, stdout_fd)
+            } else {
+                (-1, -1)
+            };
+            self.pollfds[i * 2] = PollFd::new(stderr_fd, PollFlags::POLLIN);
+            self.pollfds[i * 2 + 1] = PollFd::new(stdout_fd, PollFlags::POLLIN);
+        }
+        self.pollfds[self.snare.config.maxjobs * 2] =
+            PollFd::new(self.snare.event_read_fd, PollFlags::POLLIN);
+    }
+
+    /// If the user has specified an email address, send the contents of
+    fn sendemail(&self, mut file: &File) {
+        if let Some(ref toaddr) = self.snare.config.email {
+            let mut buf = Vec::new();
+            buf.extend("Subject: snare error\n\n".as_bytes());
+            file.seek(SeekFrom::Start(0)).ok();
+            file.read_to_end(&mut buf).ok();
+
+            let fromea = match EmailAddress::new(format!("{}@{}", username(), hostname())) {
+                Ok(ea) => Some(ea),
+                Err(_) => None,
+            };
+
+            match EmailAddress::new(toaddr.to_string()) {
+                Ok(toea) => {
+                    let email = SendableEmail::new(
+                        Envelope::new(fromea, vec![toea]).unwrap(),
+                        "na".to_string(),
+                        buf,
+                    );
+
+                    let mut sender = SendmailTransport::new();
+                    if let Err(e) = sender.send(email) {
+                        self.warning(&format!("Couldn't send email: {:?}", e));
+                    }
+                }
+                Err(_) => self.warning(&format!("Invalid To: email address {}", toaddr)),
+            }
+        }
+    }
+
+    /// Log a warning to the user. This might be to stderr, a log file, or email.
+    fn warning(&self, msg: &str) {
+        eprintln!("Warning: {}\n", msg);
+    }
+}
+
+struct Job {
+    child: Child,
+    /// This TempDir will be dropped, and its file system contents removed, when this Job is dropped.
+    _tempdir: TempDir,
```

Comment:
So is `_tempdir` needed? The underscore suggests it isn't used?

Reply:
It's so that it's kept alive until `Job` is dropped.